web-api/src/storage/elastic/client.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, fmt::Debug, str::FromStr, sync::Arc};
+
+use reqwest::{
+    header::{HeaderMap, HeaderValue, CONTENT_TYPE},
+    Body,
+    StatusCode,
+    Url,
+};
+use secrecy::{ExposeSecret, Secret};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
+use tracing::error;
+
+use crate::{
+    app::SetupError,
+    error::common::InternalError,
+    utils::{serialize_redacted, serialize_to_ndjson},
+    Error,
+};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(default)]
+pub(crate) struct Config {
+    url: String,
+    user: String,
+    #[serde(serialize_with = "serialize_redacted")]
+    password: Secret<String>,
+    index_name: String,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            url: "http://localhost:9200".into(),
+            user: "elastic".into(),
+            password: String::from("changeme").into(),
+            index_name: "test_index".into(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct Client {
+    config: Arc<Config>,
+    url_to_index: Url,
+    client: reqwest::Client,
+}
+
+impl Client {
+    pub(crate) fn builder(config: &Config) -> Result<ClientBuilder, SetupError> {
+        Ok(ClientBuilder {
+            config: Arc::new(config.clone()),
+            base_url: Arc::new(config.url.parse::<SegmentableUrl>()?),
+            client: reqwest::Client::new(),
+        })
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct ClientBuilder {
+    config: Arc<Config>,
+    base_url: Arc<SegmentableUrl>,
+    client: reqwest::Client,
+}
+
+impl ClientBuilder {
+    pub(crate) fn build(&self) -> Client {
+        Client {
+            config: self.config.clone(),
+            url_to_index: self
+                .base_url
+                .with_segments([&self.config.index_name])
+                .into(),
+            client: self.client.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
```

Comment:
This is mostly just moved except that:

- we a generic over `DocumentId` (it will be moved to `web-api-shard` where there is no document id
- has a custom error type instead of `InternalError` (**to be done**)